controller/work_item_labels_blackbox_test.go

```diff
@@ -0,0 +1,184 @@
+package controller_test
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/fabric8-services/fabric8-wit/account"
+	"github.com/fabric8-services/fabric8-wit/app"
+	"github.com/fabric8-services/fabric8-wit/app/test"
+	. "github.com/fabric8-services/fabric8-wit/controller"
+	"github.com/fabric8-services/fabric8-wit/gormapplication"
+	"github.com/fabric8-services/fabric8-wit/gormsupport/cleaner"
+	"github.com/fabric8-services/fabric8-wit/gormtestsupport"
+	"github.com/fabric8-services/fabric8-wit/label"
+	"github.com/fabric8-services/fabric8-wit/resource"
+	testsupport "github.com/fabric8-services/fabric8-wit/test"
+	tf "github.com/fabric8-services/fabric8-wit/test/testfixture"
+	wittoken "github.com/fabric8-services/fabric8-wit/token"
+	"github.com/goadesign/goa"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type TestWorkItemLabelREST struct {
+	gormtestsupport.DBTestSuite
+	db           *gormapplication.GormDB
+	ctx          context.Context
+	clean        func()
+	testIdentity account.Identity
+}
+
+func TestRunWorkItemLabelREST(t *testing.T) {
+	resource.Require(t, resource.Database)
+	suite.Run(t, &TestWorkItemLabelREST{DBTestSuite: gormtestsupport.NewDBTestSuite("../config.yaml")})
+}
+
+func (l *TestWorkItemLabelREST) SetupTest() {
+	l.db = gormapplication.NewGormDB(l.DB)
+	l.clean = cleaner.DeleteCreatedEntities(l.DB)
+	req := &http.Request{Host: "localhost"}
+	params := url.Values{}
+	l.ctx = goa.NewContext(context.Background(), nil, req, params)
+	fixture := tf.NewTestFixture(l.T(), l.DB, tf.Identities(1))
+	require.NotNil(l.T(), fixture)
+	require.Len(l.T(), fixture.Identities, 1)
+	l.testIdentity = *fixture.Identities[0]
+}
+
+func (l *TestWorkItemLabelREST) TearDownTest() {
+	l.clean()
+}
+
+func (l *TestWorkItemLabelREST) SecuredController() (*goa.Service, app.WorkitemController) {
+	pub, _ := wittoken.RSAPublicKey()
+	svc := testsupport.ServiceAsUser("WorkItemLabel-Service", wittoken.NewManager(pub), l.testIdentity)
+	return svc, NewWorkitemController(svc, l.db, l.Configuration)
+}
+
+func (l *TestWorkItemLabelREST) UnSecuredController() (*goa.Service, app.WorkitemController) {
+	svc := goa.New("WorkItemLabel-Service")
+	return svc, NewWorkitemController(svc, l.db, l.Configuration)
+}
+
+func (l *TestWorkItemLabelREST) TestAttachDetachLabelToWI() {
+	wiCnt := 2
+	lblCnt := 3
+	fixtures := tf.NewTestFixture(l.T(), l.DB, tf.Spaces(1), tf.Iterations(1), tf.Areas(1), tf.WorkItems(wiCnt), tf.Labels(lblCnt))
```

Comment:
I think creating labels will automatically create other entities.

Reply:
Good point @baijum, you are right, I had a discussion with @kwk on the same.
While I was patching the WI with labels, logic at this place gets executed https://github.com/fabric8-services/fabric8-wit/blob/master/controller/workitem.go#L263 which is trying to read RootIteration/RootArea which are assumed to be created for every space.
Using TestFixtures, we are not creating Root* things for space. So root iteration and area are not available when I just do `tf.NewTestFixture(l.T(), l.DB, tf.WorkItems(wiCnt), tf.Labels(lblCnt))`.

Reply:
Is it something missing in the TextFixture code, then ?

Reply:
If one creates a space using repo, only space is created.
If one creates a space using conroller(HTTP POST) then a space, rootIteration, rootArea is created.
Our TestFixtures are for repo functions, so they do not create Root* entries... If we have controller level fixtures then we can have Root* entries without thinking about it..

Reply:
hum, my understanding was that the TextFixture would create all required entities (`space`, etc.) if you ask for work items or labels.